federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,876 @@
+	}
+	hpa2.Spec.MaxReplicas = hpa1.Spec.MaxReplicas
+	return fedutil.ObjectMetaAndSpecEquivalent(hpa1, hpa2)
+}
+
+type replicaNums struct {
+	min int32
+	max int32
+}
+
+type hpaFederatedStatus struct {
+	lastScaleTime *metav1.Time
+	// Indicates how many clusters have hpa/replicas.
+	// Used to average the cpu utilization which is
+	// reflected to the federation user.
+	count                           int32
+	currentCPUUtilizationPercentage *int32
+	currentReplicas                 int32
+	desiredReplicas                 int32
+}
+
+type hpaSchedulingInfo struct {
+	scheduleState map[string]*replicaNums
+	fedStatus     hpaFederatedStatus
+}
+
+// List of cluster names.
+type hpaLists struct {
+	// Stores names of those clusters which can offer min.
+	availableMin sets.String
+	// Stores names of those clusters which can offer max.
+	availableMax sets.String
+	// Stores names of those clusters which do not have hpa yet.
+	noHpa sets.String
+}
+
+func (a *HpaAdapter) GetSchedule(obj pkgruntime.Object, key string, clusters []*federationapi.Cluster, informer fedutil.FederatedInformer) (interface{}, error) {
+	currentClusterObjs, err := getCurrentClusterObjs(informer, key, clusters)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialise averaged cpu utilisation for this reconcile.
+	var ccup int32 = 0
+	fedStatus := hpaFederatedStatus{
+		currentCPUUtilizationPercentage: &ccup,
+		count:           int32(0),
+		desiredReplicas: int32(0),
+		currentReplicas: int32(0),
+	}
+	fedHpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	// We assign the last known scale time here, which we update with
+	// the latest time from among all clusters in ScheduleObject()
+	if fedHpa.Status.LastScaleTime != nil {
+		t := metav1.NewTime(fedHpa.Status.LastScaleTime.Time)
+		fedStatus.lastScaleTime = &t
+	}
+
+	return &hpaSchedulingInfo{
+		scheduleState: getHpaScheduleState(obj, currentClusterObjs),
+		fedStatus:     fedStatus,
+	}, nil
+}
+
+func getCurrentClusterObjs(informer fedutil.FederatedInformer, key string, clusters []*federationapi.Cluster) (map[string]pkgruntime.Object, error) {
+	currentClusterObjs := make(map[string]pkgruntime.Object)
+	for _, cluster := range clusters {
+		clusterName := cluster.Name
+		clusterObj, found, err := informer.GetTargetStore().GetByKey(clusterName, key)
+		if err != nil {
+			return nil, err
+		}
+		currentClusterObjs[clusterName] = nil
+		if found {
+			currentClusterObjs[clusterName] = clusterObj.(pkgruntime.Object)
+		}
+	}
+	return currentClusterObjs, nil
+}
+
+func getHpaScheduleState(fedObj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) map[string]*replicaNums {
+	fedHpa := fedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	requestedMin := int32(1)
+	if fedHpa.Spec.MinReplicas != nil {
+		requestedMin = *fedHpa.Spec.MinReplicas
+	}
+	requestedReplicas := replicaNums{
+		min: requestedMin,
+		max: fedHpa.Spec.MaxReplicas,
+	}
+	// replica distribution count, per cluster
+	rdc := replicaNums{
```

Comment:
Fair enough.  I was thinking of what the desired initial placement of replicas might be before the rebalancing starts to happen.  But I agree, that requires some more thought and discussion, and can be done as a followup.

In my experience, a user might know approximately what the relative weightings of clusters are (e.g. us:eu:asia = 3:2:1) and generally that's how they want their min and max replicas distributed in the normal case.  It's only in relatively unusual  situations like failures or unexpected traffic shifts that they want to redistribute max replicas (e.g. SuperBowl happens in the U.S, while Europe is asleep, so they want some unneeded "max" capacity in Europe moved to the U.S. to prevent the U.S from overloading, rather than have it hit it's max per it's originally specified weight).

Happy to defer to followup discussion/PR.